burn-dataset/src/vision/image_folder.rs

```diff
@@ -0,0 +1,382 @@
+                .map(|&x| PixelDepth::U16(x))
+                .collect(),
+            ColorType::Rgb32F => image
+                .into_rgb32f()
+                .iter()
+                .map(|&x| PixelDepth::F32(x))
+                .collect(),
+            ColorType::Rgba32F => image
+                .into_rgba32f()
+                .iter()
+                .map(|&x| PixelDepth::F32(x))
+                .collect(),
+            _ => panic!("Unrecognized image color type"),
+        };
+
+        ImageDatasetItem {
+            image: img_vec,
+            target: ImageTarget::Label(*label),
+        }
+    }
+}
+
+type ClassificationDatasetMapper = MapperDataset<
+    InMemDataset<ImageDatasetItemRaw>,
+    PathToImageClassificationItem,
+    ImageDatasetItemRaw,
+>;
+
+/// A generic dataset to load classification images from disk.
+pub struct ImageFolderDataset {
+    dataset: ClassificationDatasetMapper,
+}
+
+impl Dataset<ImageDatasetItem> for ImageFolderDataset {
+    fn get(&self, index: usize) -> Option<ImageDatasetItem> {
+        self.dataset.get(index)
+    }
+
+    fn len(&self) -> usize {
+        self.dataset.len()
+    }
+}
+
+impl ImageFolderDataset {
+    /// Create an image classification dataset from the root folder.
+    ///
+    /// # Arguments
+    ///
+    /// * `root` - Dataset root folder.
+    ///
+    /// # Returns
+    /// A new dataset instance.
+    pub fn new<P: AsRef<Path>>(root: P) -> Self {
+        // New dataset containing any of the supported file types
+        ImageFolderDataset::new_with(root, &SUPPORTED_FILES)
+    }
+
+    /// Create an image classification dataset from the root folder.
+    /// The included images are filtered based on the provided extensions.
+    ///
+    /// # Arguments
+    ///
+    /// * `root` - Dataset root folder.
+    /// * `extensions` - List of allowed extensions.
+    ///
+    /// # Returns
+    /// A new dataset instance.
+    pub fn new_with<P, S>(root: P, extensions: &[S]) -> Self
+    where
+        P: AsRef<Path>,
+        S: AsRef<str>,
+    {
+        fn check_extension<S: AsRef<str>>(extension: &S) -> String {
+            let extension = extension.as_ref();
+            assert!(
+                SUPPORTED_FILES.contains(&extension),
+                "Invalid extension '{}'",
+                extension
+            );
+
+            extension.to_string()
+        }
+        // Glob all images with extensions
+        let walker = globwalk::GlobWalkerBuilder::from_patterns(
+            root.as_ref(),
+            &[format!(
+                "*.{{{}}}", // "*.{ext1,ext2,ext3}
+                extensions
+                    .iter()
+                    .map(check_extension)
+                    .collect::<Vec<_>>()
+                    .join(",")
+            )],
+        )
+        .follow_links(true)
+        .sort_by(|p1: &DirEntry, p2: &DirEntry| p1.path().cmp(p2.path())) // order by path
+        .build()
+        .unwrap()
+        .filter_map(Result::ok);
```

Comment:
I believe we should return a `Result` from the dataset creation. Similar to the Hugging Face dataset loader, certain errors may occur that are not programmer errors, such as file disk access issues. Therefore, implementing error handling in this section is a good practice.

Reply:
That's the type of feedback I was looking for 🙏 I am still not always sure when it is better practice to use `unwrap()` or other such methods vs returning a `Result`.

Reply:
`unwrap` => dev errors (made by us or the user such as dimension mismatch in the tensor API) 

`runtime error` => Result